src/ast/builder.rs

```diff
@@ -322,6 +342,95 @@ where
     }
 }
 
+/// A builder that can produce an `Implication` around the results of two inner builders.
+#[derive(Debug)]
+pub struct CompleteImplies<L, R> {
+    left: L,
+    right: R,
+}
+
+impl<L, R> CompleteImplies<L, R> {
+    /// Create a new `CompleteImplies` builder.
+    pub fn new(left: L, right: R) -> Self {
+        Self { left, right }
+    }
+
+    /// Append an 'And' to the sentence.
+    pub fn and(self) -> IncompleteAnd<IncompleteImplies<Builder, L>, R> {
+        IncompleteAnd::new(IncompleteImplies::new(Builder, self.left), self.right)
+    }
+
+    /// Append an `Or` to the sentence.
+    pub fn or(self) -> IncompleteOr<IncompleteImplies<Builder, L>, R> {
+        IncompleteOr::new(IncompleteImplies::new(Builder, self.left), self.right)
+    }
+
+    /// Append an `Implication` to the sentence.
+    pub fn implies(self) -> IncompleteImplies<Builder, Self> {
+        IncompleteImplies::new(Builder, self)
```

Comment:
Careful with this one - implies is not associative: **((A -> B) -> C) != (A -> (B -> C))**

| A | B | C | A -> B | B -> C | (A -> B) -> C | A -> (B -> C) |
| -- | - | -- | ------- | ------- | -------------- | -------------- |
| F  | F | F  | T         | T         | **F**              | **T**              |
| F  | F | T  | T         | T         | **T**              | **T**              |
| F  | T | F  | T         | F         | **F**              | **T**              |
| F  | T | T  | T         | T         | **T**              | **T**              |
| T  | F | F  | F         | T         | **T**              | **T**              |
| T  | F | T  | F         | T         | **T**              | **T**              |
| T  | T | F  | T         | F         | **F**              | **F**              |
| T  | T | T  | T         | T         | **T**              | **T**              |

Logical operators are also right-associative: **(A -> B -> C) == (A -> (B -> C))**

```suggestion
        IncompleteImplies::new(IncompleteImplies::new(Builder, self.left), self.right)
```

This will setup the builder to create an implies like: **self.left -> (self.right -> ?)**

Reply:
Ooh it's probably a good idea to make a test to make sure the associativity is correct, i.e.,

```rust
let expected = implies(atom("a"), implies(atom("b"), atom("c")));
let actual = Builder.atom("a").implies().atom("b").implies().atom("c").build();

assert_eq!(expected, actual);
```